pkgdb/src/resolver/environment.cc

```diff
@@ -457,22 +458,99 @@ Environment::getGroupInput( const InstallDescriptors & group,
 }
 
 
+/* -------------------------------------------------------------------------- */
+
+std::variant<InstallID, SystemPackages>
+Environment::tryResolveGroupIn( const InstallDescriptors & group,
+                                const pkgdb::PkgDbInput &  input,
+                                const System &             system )
+{
+  std::unordered_map<InstallID, std::optional<pkgdb::row_id>> pkgRows;
+
+  /* Loop over each descriptor. */
+  for ( const auto & [iid, descriptor] : group )
+    {
+      /* Skip unrequested systems. */
+      if ( descriptor.systems.has_value()
+           && ( std::find( descriptor.systems->begin(),
+                           descriptor.systems->end(),
+                           system )
+                == descriptor.systems->end() ) )
+        {
+          pkgRows.emplace( iid, std::nullopt );
+          continue;
+        }
+
+      /* Try resolving.
+       * If resolution fails, return the _iid_ of the **first**
+       * failed descriptor. */
+      std::optional<pkgdb::row_id> maybeRow
+        = this->tryResolveDescriptorIn( descriptor, input, system );
+      if ( maybeRow.has_value() || descriptor.optional )
+        {
+          pkgRows.emplace( iid, maybeRow );
+        }
+      else { return iid; }
+    }
+
+  /* Convert to `LockedPackageRaw's */
+  SystemPackages pkgs;
+  LockedInputRaw lockedInput( input );
+  auto           dbRO = input.getDbReadOnly();
+  for ( const auto & [iid, maybeRow] : pkgRows )
+    {
+      if ( maybeRow.has_value() )
+        {
+          pkgs.emplace( iid,
+                        Environment::lockPackage( lockedInput,
+                                                  *dbRO,
+                                                  *maybeRow,
+                                                  group.at( iid ).priority ) );
+        }
+      else { pkgs.emplace( iid, std::nullopt ); }
+    }
+
+  return pkgs;
+}
+
+
+/* -------------------------------------------------------------------------- */
+
+/**
+ * @brief Extract the name of a group from a set of descriptors, or "default"
+ *        if no descriptors declare a `packageGroup`.
+ */
+[[nodiscard]] static inline const std::string &
+getGroupName( const InstallDescriptors & group )
+{
+  if ( const auto & descriptor = group.begin();
+       ( descriptor != group.end() ) && descriptor->second.group.has_value() )
+    {
+      return *descriptor->second.group;
+    }
+  static const std::string defaultName = "default";
+  return defaultName;
+}
+
+
 /* -------------------------------------------------------------------------- */
 
 ResolutionResult
 Environment::tryResolveGroup( const InstallDescriptors & group,
-                              const System &             system )
+                              const System &             system,
+                              bool                       ignoreOld )
```

Comment:
nonblocking: do we need `ignoreOld` anymore? Unless you're expecting to use it in the future, we could drop it